core/handler_integration_test.go

```diff
@@ -0,0 +1,118 @@
+	// We generally desire messages to expire from the pool before the queue so that retries are
+	// accepted.
+	maxAge := uint(10)
+	gasPrice := types.NewGasPrice(1)
+	gasUnits := types.NewGasUnits(1000)
+
+	makeHandler := func(provider *core.FakeProvider, root types.TipSet) *core.NewHeadHandler {
+		mpool := core.NewMessagePool(config.NewDefaultConfig().Mpool, th.NewMockMessagePoolValidator())
+		inbox := core.NewInbox(mpool, maxAge, provider, provider)
+		queue := core.NewMessageQueue()
+		publisher := core.NewDefaultMessagePublisher(&core.MockNetworkPublisher{}, "Topic", mpool)
+		policy := core.NewMessageQueuePolicy(provider, maxAge)
+		outbox := core.NewOutbox(signer, &core.FakeValidator{}, queue, publisher, policy, provider, provider)
+
+		return core.NewHandler(inbox, outbox, provider, root)
+	}
+
+	t.Run("test send after reverted message", func(t *testing.T) {
+		provider := core.NewFakeProvider(t)
+		root := provider.NewGenesis()
+		actr, _ := account.NewActor(types.ZeroAttoFIL)
+		actr.Nonce = 42
+		provider.SetHeadAndActor(t, root.Key(), sender, actr)
+
+		handler := makeHandler(provider, root)
+		outbox := handler.Outbox
+		inbox := handler.Inbox
+
+		// First, send a message and expect to find it in the message queue and pool.
+		mid1, err := outbox.Send(ctx, sender, dest, types.ZeroAttoFIL, gasPrice, gasUnits, true, "method1")
+		require.NoError(t, err)
+		require.Equal(t, 1, len(outbox.Queue().List(sender))) // Message is in the queue.
+		msg1, found := inbox.Pool().Get(mid1)
+		require.True(t, found) // Message is in the pool.
+		assert.True(t, msg1.Equals(outbox.Queue().List(sender)[0].Msg))
+
+		// Receive the message in a block.
+		left := provider.BuildOneOn(root, func(b *chain.BlockBuilder) {
+			b.AddMessages([]*types.SignedMessage{msg1}, types.EmptyReceipts(1))
+		})
+		require.NoError(t, handler.HandleNewHead(ctx, left))
+		assert.Equal(t, 0, len(outbox.Queue().List(sender))) // Gone from queue.
+		_, found = inbox.Pool().Get(mid1)
+		assert.False(t, found) // Gone from pool.
+
+		// Now re-org the chain to un-mine that message.
+		right := provider.BuildOneOn(root, func(b *chain.BlockBuilder) {
+			// No messages.
+		})
+		require.NoError(t, handler.HandleNewHead(ctx, right))
+		assert.Equal(t, 0, len(outbox.Queue().List(sender))) // Message does not return to queue.
+		_, found = inbox.Pool().Get(mid1)
+		assert.True(t, found) // Message returns to pool to be mined again.
+
+		// Send another message from the same account.
+		// First, send a message and expect to find it in the message queue and pool.
+		_, err = outbox.Send(ctx, sender, dest, types.ZeroAttoFIL, gasPrice, gasUnits, true, "method2")
+		// This case causes the nonce to be wrongly calculated, since the first, now-unmined message
+		// is not in the outbox, and actor state has not updated, but the message pool already has
+		// a message with the same nonce.
+		assert.Error(t, err, "Fix this in issue #3052")
```

Comment:
This is confusing me. The "Fix this..." message will appear in the test output if there is not an error. It appears that this test is reproducing #3052. Maybe move the "Fix this..." to a comment?

Reply:
Also consider asserting the error is a "same actor and nonce" error to demonstrate it's actually reproducing the problem.

Reply:
I changed the message to `"#3052 is fixed!"`